opacus/optimizers/optimizer.py

```diff
@@ -1,13 +1,38 @@
 from __future__ import annotations
 
-from typing import Callable, List, Optional
+from typing import Callable, List, Optional, Union
 
 import torch
 from opacus.optimizers.utils import params
 from torch import nn
 from torch.optim import Optimizer
 
 
+def _mark_as_processed(obj: Union[torch.Tensor, List[torch.Tensor]]):
+    if isinstance(obj, torch.Tensor):
+        obj._processed = True
+    elif isinstance(obj, list):
+        for x in obj:
+            x._processed = True
+
+
+def _check_processed_flag_tensor(x: torch.Tensor):
+    if hasattr(x, "_processed"):
+        raise ValueError(
+            "Gradients hasn't been cleared since the last optimizer step. "
+            "In order to obtain privacy guarantees you must call optimizer.zero_grad()"
+            "on each step"
+        )
```

Comment:
Check my understanding please: `zero_grad` not only changes `.grad`, but also clears all the subfields? Otherwise, how it works? In [the docs](https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.zero_grad.html) it is said that `zero_grad` only changes `.grad`.

Reply:
we override `zero_grad` method in DPOptimizer and GradSampleModule to also clear fields created and managed by opacus (i.e. `grad_sample` and `summed_grad`)